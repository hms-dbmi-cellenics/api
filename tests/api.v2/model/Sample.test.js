// @ts-nocheck
const _ = require('lodash');

const { mockSqlClient, mockTrx } = require('../mocks/getMockSqlClient')();

const Sample = require('../../../src/api.v2/model/Sample');

const getSamplesResponse = require('../mocks/data/getSamplesResponse.json');

const mockExperimentId = 'mockExperimentId';
// @ts-nocheck
// Disabled ts because it doesn't recognize jest mocks
const tableNames = require('../../../src/api.v2/model/tableNames');

jest.mock('../../../src/sql/sqlClient', () => ({
  get: jest.fn(() => mockSqlClient),
}));

jest.mock('../../../src/sql/helpers', () => ({
  collapseKeysIntoObject: jest.fn(),
  collapseKeyIntoArray: jest.fn(),
  replaceNullsWithObject: jest.requireActual('../../../src/sql/helpers').replaceNullsWithObject,
}));

const mockSampleId = 'mockSampleId';
const mockSampleFileId = 'sampleFileId';
const mockSampleFileType = 'features10x';
const mockOverwriteExisting = true;


describe('model/Sample', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Get samples works correctly', async () => {
    await new Sample().getSamples(mockExperimentId).toString();
    expect(mockSqlClient.where.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.from.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.select.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.join.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.groupBy.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.raw.mock.calls).toMatchSnapshot();
    expect(mockSqlClient.leftJoin.mock.calls).toMatchSnapshot();
  });

  it('setNewFile works correctly if valid params are passed', async () => {
    mockTrx.ref.mockImplementationOnce(() => 'sf_mapsample_file_idRef');

    await new Sample().setNewFile(mockSampleId, mockSampleFileId, mockSampleFileType, mockOverwriteExisting);

    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE_TO_SAMPLE_FILE_MAP);

    // Test delete from SAMPLE_TO_SAMPLE_FILE_MAP query
    expect(mockTrx.from).toHaveBeenCalledWith({ sf_map: tableNames.SAMPLE_TO_SAMPLE_FILE_MAP });
    expect(mockTrx.where).toHaveBeenCalledWith({ sample_id: mockSampleId });
    expect(mockTrx.andWhere).toHaveBeenCalledWith('sample_file_id', '=', expect.anything());

    expect(mockTrx.select).toHaveBeenCalledWith(['id']);
    expect(mockTrx.from).toHaveBeenCalledWith({ sf: tableNames.SAMPLE_FILE });
    expect(mockTrx.where).toHaveBeenCalledWith('sf.id', '=', 'sf_mapsample_file_idRef');
    expect(mockTrx.andWhere).toHaveBeenCalledWith('sf.sample_file_type', '=', mockSampleFileType);

    // Test add new sample file reference
    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE_TO_SAMPLE_FILE_MAP);
    expect(mockTrx.insert).toHaveBeenCalledWith({
      sample_id: mockSampleId,
      sample_file_id: mockSampleFileId,
    });
  });

  it('copyTo works correctly', async () => {
    const fromExperimentId = 'fromExperimentIdMock';
    const toExperimentId = 'toExperimentIdMock';
    const samplesOrder = Object.values(getSamplesResponse).map((sample) => sample.id);

    mockTrx.returning.mockImplementationOnce(() => Promise.resolve([{ id: 0, key: 'Track0' }]));

    const getSamplesSpy = jest.spyOn(Sample.prototype, 'getSamples')
      .mockImplementationOnce(() => Promise.resolve(getSamplesResponse));


    await new Sample().copyTo(fromExperimentId, toExperimentId, samplesOrder);


    expect(getSamplesSpy).toHaveBeenCalledWith(fromExperimentId);

    expect(mockTrx).toHaveBeenCalledWith(tableNames.METADATA_TRACK);
    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE);
    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE_IN_METADATA_TRACK_MAP);
    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE_TO_SAMPLE_FILE_MAP);

    expect(mockTrx.insert.mock.calls).toMatchSnapshot();
    expect(mockTrx.returning.mock.calls).toMatchSnapshot();
  });

  it('copyTo works correctly if experiment has no metadata tracks', async () => {
    const fromExperimentId = 'fromExperimentIdMock';
    const toExperimentId = 'toExperimentIdMock';

    const noMetadataGetSamplesResponse = _.cloneDeep(getSamplesResponse);
    noMetadataGetSamplesResponse.forEach((sample) => {
      // eslint-disable-next-line no-param-reassign
      sample.metadata = {};
    });

    const samplesOrder = Object.values(noMetadataGetSamplesResponse).map((sample) => sample.id);

    mockTrx.returning.mockImplementationOnce(() => Promise.resolve([{ id: 0, key: 'Track0' }]));

    const getSamplesSpy = jest.spyOn(Sample.prototype, 'getSamples')
      .mockImplementationOnce(() => Promise.resolve(noMetadataGetSamplesResponse));


    await new Sample().copyTo(fromExperimentId, toExperimentId, samplesOrder);


    expect(getSamplesSpy).toHaveBeenCalledWith(fromExperimentId);

    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE);
    expect(mockTrx).toHaveBeenCalledWith(tableNames.SAMPLE_TO_SAMPLE_FILE_MAP);
    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.METADATA_TRACK);
    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.SAMPLE_IN_METADATA_TRACK_MAP);

    expect(mockTrx.insert.mock.calls).toMatchSnapshot();
    expect(mockTrx.returning.mock.calls).toMatchSnapshot();
  });

  it('copyTo works correctly if experiment has no samples', async () => {
    const fromExperimentId = 'fromExperimentIdMock';
    const toExperimentId = 'toExperimentIdMock';

    const samplesOrder = [];

    const copiedSamplesOrder = await new Sample().copyTo(
      fromExperimentId, toExperimentId, samplesOrder,
    );

    expect(copiedSamplesOrder).toHaveLength(0);

    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.SAMPLE);
    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.SAMPLE_TO_SAMPLE_FILE_MAP);
    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.METADATA_TRACK);
    expect(mockTrx).not.toHaveBeenCalledWith(tableNames.SAMPLE_IN_METADATA_TRACK_MAP);
  });

  it('updateOption works correctly', async () => {
    const mockOptions = { someOption: true, otherOption: false };
    const JSONBString = `'${JSON.stringify(mockOptions)}'::jsonb`;

    mockSqlClient.raw.mockImplementationOnce((x) => x);

    await new Sample().updateOption(mockExperimentId, mockOptions);

    expect(mockSqlClient.update());

    expect(mockSqlClient.raw.mock.calls[0][0]).toEqual(JSONBString);
    expect(mockSqlClient.update.mock.calls[0][0]).toEqual({ options: JSONBString });
    expect(mockSqlClient.where.mock.calls[0][0]).toEqual({ experiment_id: mockExperimentId });
  });
});
