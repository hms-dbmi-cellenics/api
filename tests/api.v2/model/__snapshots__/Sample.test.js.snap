// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model/Sample Get samples works correctly 1`] = `
Array [
  Array [
    "s.experiment_id",
    "mockExperimentId",
  ],
  Array [
    "s.experiment_id",
    "mockExperimentId",
  ],
]
`;

exports[`model/Sample Get samples works correctly 2`] = `
Array [
  Array [
    Object {
      "s": "sample",
    },
  ],
  Array [
    [MockFunction],
  ],
  Array [
    Object {
      "s": "sample",
    },
  ],
  Array [
    [MockFunction],
  ],
  Array [
    [MockFunction],
  ],
]
`;

exports[`model/Sample Get samples works correctly 3`] = `
Array [
  Array [
    Array [
      "id",
      "experiment_id",
      "name",
      "sample_technology",
      "valid",
      "validation_message",
      "created_at",
      "updated_at",
      [MockFunction],
    ],
  ],
  Array [
    Array [
      "s.id",
      "s.experiment_id",
      "s.name",
      "s.sample_technology",
      "s.valid",
      "s.validation_message",
      "s.created_at",
      "s.updated_at",
      "m.key",
      "sm_map.value",
    ],
  ],
  Array [
    Array [
      "id",
      [MockFunction],
    ],
  ],
  Array [
    Array [
      "sf.sample_file_type",
      "sf.size",
      "sf.upload_status",
      "sf.s3_path",
      "sf.id as sf_id",
      "s.id",
    ],
  ],
  Array [
    "*",
  ],
]
`;

exports[`model/Sample Get samples works correctly 4`] = `
Array [
  Array [
    "sample_to_sample_file_map as sf_map",
    "s.id",
    "sf_map.sample_id",
  ],
  Array [
    "sample_file as sf",
    "sf.id",
    "sf_map.sample_file_id",
  ],
  Array [
    [MockFunction],
    "select_metadata.id",
    "select_sample_file.id",
  ],
]
`;

exports[`model/Sample Get samples works correctly 5`] = `
Array [
  Array [
    Array [
      "id",
      "experiment_id",
      "name",
      "sample_technology",
      "valid",
      "validation_message",
      "created_at",
      "updated_at",
    ],
  ],
  Array [
    "id",
  ],
]
`;

exports[`model/Sample Get samples works correctly 6`] = `
Array [
  Array [
    "undefined as metadata",
  ],
  Array [
    "jsonb_object_agg(sample_file_type,json_build_object('sample_file_type',sample_file_type,'size',size,'upload_status',upload_status,'s3_path',s3_path,'id',sf_id)) as files",
  ],
]
`;

exports[`model/Sample Get samples works correctly 7`] = `
Array [
  Array [
    "metadata_track as m",
    "s.experiment_id",
    "m.experiment_id",
  ],
  Array [
    "sample_in_metadata_track_map as sm_map",
    Object {
      "m.id": "sm_map.metadata_track_id",
      "s.id": "sm_map.sample_id",
    },
  ],
]
`;

exports[`model/Sample copyTo works correctly 1`] = `
Array [
  Array [
    Array [
      Object {
        "experiment_id": "toExperimentIdMock",
        "key": "Track0",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "experiment_id": "toExperimentIdMock",
        "id": "mock-uuid",
        "name": "KO",
        "sample_technology": "10x",
      },
      Object {
        "experiment_id": "toExperimentIdMock",
        "id": "mock-uuid",
        "name": "WT1",
        "sample_technology": "10x",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "sample_file_id": "2afe253f-5542-4a90-9fc9-17acdc3ecaf1",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "59b8221d-dd3f-4bb9-ad9d-1166471e7143",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "415323cb-72dc-4f42-b452-c12422ea042f",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "d5c957b9-a6cb-4828-8ee6-1f9da8cacb6a",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "d52a84a8-4669-41b3-a8d1-38e16ddbf522",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "5b8ec36c-0f06-4a65-b924-da5189c89b7b",
        "sample_id": "mock-uuid",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "metadata_track_id": 0,
        "sample_id": "mock-uuid",
        "value": "N.A.1",
      },
      Object {
        "metadata_track_id": 0,
        "sample_id": "mock-uuid",
        "value": "N.A.",
      },
    ],
  ],
]
`;

exports[`model/Sample copyTo works correctly 2`] = `
Array [
  Array [
    Array [
      "id",
      "key",
    ],
  ],
]
`;

exports[`model/Sample copyTo works correctly if experiment has no metadata tracks 1`] = `
Array [
  Array [
    Array [
      Object {
        "experiment_id": "toExperimentIdMock",
        "id": "mock-uuid",
        "name": "KO",
        "sample_technology": "10x",
      },
      Object {
        "experiment_id": "toExperimentIdMock",
        "id": "mock-uuid",
        "name": "WT1",
        "sample_technology": "10x",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "sample_file_id": "2afe253f-5542-4a90-9fc9-17acdc3ecaf1",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "59b8221d-dd3f-4bb9-ad9d-1166471e7143",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "415323cb-72dc-4f42-b452-c12422ea042f",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "d5c957b9-a6cb-4828-8ee6-1f9da8cacb6a",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "d52a84a8-4669-41b3-a8d1-38e16ddbf522",
        "sample_id": "mock-uuid",
      },
      Object {
        "sample_file_id": "5b8ec36c-0f06-4a65-b924-da5189c89b7b",
        "sample_id": "mock-uuid",
      },
    ],
  ],
]
`;

exports[`model/Sample copyTo works correctly if experiment has no metadata tracks 2`] = `Array []`;
